src/fable-library/List.fs

```diff
@@ -6,153 +6,185 @@ module List
+    for i = xs.Count - 1 downto 0 do
+        acc <- folder acc xs.Values.[i]
+    acc
+
+let foldBack (folder: 'T -> 'acc -> 'acc) (xs: 'T list) (state: 'acc) =
+    let mutable acc = state
+    for i = 0 to xs.Count - 1 do
+        acc <- folder xs.Values.[i] acc
+    acc
+
+let reverse (xs: 'a list) =
+    fold (fun acc x -> cons x acc) List.Empty xs
 
 let toSeq (xs: 'a list): 'a seq =
     Seq.map id xs
 
 let ofSeq (xs: 'a seq): 'a list =
-    Seq.fold (fun acc x -> x::acc) [] xs
+    Seq.fold (fun acc x -> cons x acc) List.Empty xs
     |> reverse
 
 let concat (lists: seq<'a list>) =
-    Seq.fold (fold (fun acc x -> x::acc)) [] lists
+    Seq.fold (fold (fun acc x -> cons x acc)) List.Empty lists
     |> reverse
 
-let rec foldIndexed2Aux f i acc bs cs =
-    match bs, cs with
-    | [], [] -> acc
-    | x::xs, y::ys -> foldIndexed2Aux f (i+1) (f i acc x y) xs ys
-    | _ -> invalidOp "Lists had different lengths"
-
-let foldIndexed2<'a, 'b, 'acc> f (state: 'acc) (xs: 'a list) (ys: 'b list) =
-    foldIndexed2Aux f 0 state xs ys
-
 let fold2<'a, 'b, 'acc> f (state: 'acc) (xs: 'a list) (ys: 'b list) =
     Seq.fold2 f state xs ys
 
 let foldBack2<'a, 'b, 'acc> f (xs: 'a list) (ys: 'b list) (state: 'acc) =
     Seq.foldBack2 f xs ys state
 
-let unfold f state =
-    let rec unfoldInner acc state =
-        match f state with
-        | None -> reverse acc
-        | Some (x,state) -> unfoldInner (x::acc) state
-    unfoldInner [] state
-
-let rec foldIndexed3Aux f i acc bs cs ds =
-    match bs, cs, ds with
-    | [], [], [] -> acc
-    | x::xs, y::ys, z::zs -> foldIndexed3Aux f (i+1) (f i acc x y z) xs ys zs
-    | _ -> invalidOp "Lists had different lengths"
-
-let foldIndexed3<'a, 'b, 'c, 'acc> f (seed: 'acc) (xs: 'a list) (ys: 'b list) (zs: 'c list) =
-    foldIndexed3Aux f 0 seed xs ys zs
-
-let fold3<'a, 'b, 'c, 'acc> f (state: 'acc) (xs: 'a list) (ys: 'b list) (zs: 'c list) =
-    foldIndexed3 (fun _ acc x y z -> f acc x y z) state xs ys zs
+let unfold (gen: 'acc -> ('T * 'acc) option) (state: 'acc) =
+    let values = ResizeArray<_>()
+    let rec loop acc =
+        match gen acc with
+        | None -> ()
+        | Some (x, st) -> values.Add(x); loop st
+    loop state
+    newList values
 
 let scan<'a, 'acc> f (state: 'acc) (xs: 'a list) =
     Seq.scan f state xs |> ofSeq
 
 let scanBack<'a, 'acc> f (xs: 'a list) (state: 'acc) =
     Seq.scanBack f xs state |> ofSeq
 
-let length xs =
-    fold (fun acc _ -> acc + 1) 0 xs
-
-let append xs ys =
-    fold (fun acc x -> x::acc) ys (reverse xs)
+let append (xs: 'a list) (ys: 'a list) =
+    Seq.append xs ys |> ofSeq
 
 let collect (f: 'a -> 'b list) (xs: 'a list) =
     Seq.collect f xs |> ofSeq
 
-let map f xs =
-    fold (fun acc x -> f x::acc) [] xs
-    |> reverse
+let map f (xs: 'a list) =
+    let values = ResizeArray<_>(xs.Count)
+    for i = xs.Count - 1 downto 0 do
+        values.Add(f xs.Values.[i])
+    values.Reverse()
+    newList values
```

Comment:
Actually I just noticed we cannot compiler `ResizeArray(5)` as `new Array(5)` because this would make the array of length 5 already. We will have to emit some js if we want to allocate the array.

Reply:
```fsharp
[<Emit("new Array($0)")>]
let allocateArray (capacity: int): ResizeArray<'T> = jsNative

let values = allocateArray xs.Count
for i = 0 to xs.Count - 1 do
    values.[i] <- f xs.Values.[xs.Count - 1 - i]
```